### DISCUSSION TITLE
Advice for Language Servers using parsers written in C++

### DISCUSSION THREAD

**Discussion by gnikit** [microsoft/vscode-cpptools#9632] [Q&A] [answered]
Hi All,

We are in the process of using our own Fortran parser present in [lfortran](https://github.com/lfortran/lfortran), written in C++, as the backend of a new Language Server and I was wondering if you could share some of your expertise or at least provide some comments.
Specifically, I was wondering if you could provide some input with regards to your Language Server implementations. Reading through the [implementations](https://microsoft.github.io/language-server-protocol/implementors/servers/) in the LSP website I see that your C++ LSs are actually coded in C++. I assume that means that your parser/tokenizer is written in C++ and therefore your AST/ASR representation is also in C++. If that is true, I am curious to know if you have implemented a JSONRPC interface in C++ and then send these requests to your TypeScript client or if you have implemented some sort of interface between C++ and Typescript, like [nbind](https://github.com/charto/nbind) that is capable of translating your C++ AST/ASR to Typescript to then be converted into LS requests via the `vscode-languageserver/node` module.

Basically, implementing the LSP interface in C++, even if that is only for the server, would be a nightmare to maintain and currently there is not enough manpower to support this. So we are looking for better solutions to communicate our AST/ASR objects into Typescript. nbind, looked like a good solution but the project appears to be dead. Do you have any recommendations with regards to this?

I know that your C++ LSs are proprietary and you are limited to what you are allowed to share. If it would be preferable we can continue this conversation in private.

Thanks in advance for any and all help!

**Comment by sean-mcmanus** [comment, marked as answer]
TypeScript to LSP server communication could be done with any JSON serialization library and doesn't require RPCs. Our AST isn't sent to TypeScript -- if that's what you're trying to do then we don't have any advice on that. Our LSP server just reads the JSON messages and returns JSON results. The LSP is versioned so you can stay on an older version unless you need a newer one for a particular feature so I'm not sure what maintenance nightmare you're referring to.

**Reply by gnikit** [reply]
Thanks @sean-mcmanus for the response! Currently our approach is similar to yours albeit substantially cruder. We are still trying to figure out the design that would allow us to grow our server with the most ease; currently we are doing 2 things:
1) creating requests in C++ using JSON-RPC and sending them to a TypeScript client (not ideal since we have to write and maintain the RPC code in C++)
2) creating JSON requests in C++, send the JSON to TypeScript, read the JSON message and package it into an LSP request.

However both approaches suffer from the same problem, which is what I referred to as _maintainance nightmare_. We would have to write and maintain a complete copy (or a copy of the features the LS implements) of the LSP standard in C++ to make sure that each JSON request we send to TypeScript contains the necessary info for us to build the JSON-RPC request.
I am referring to things like [`Range`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#range), [`Position`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position), [`CompletionItemKind`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItemKind), etc. This is what I was hoping we could minimise by using an LSP SDK for C++, similar to [pygls](https://github.com/openlawlibrary/pygls#pygls) in Python but I am not aware of any such project that is actively developed and mature enough. (Hence pinning to an LSP version would not really help). 

Having gone through this once already in Python, I can say that reading the LSP standard and implementing it in language X is quite a bit of work. Do you still think that our best bet would be to go with option 2. (which is what you are doing if I understood correctly)?

**Reply by sean-mcmanus** [reply]
As you can see from our open source TypeScript, we are using the vscode-languageclient API to handle the LSP stuff (see for example https://github.com/microsoft/vscode-cpptools/pull/9524 ) and we create our client at https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/client.ts#L1363 . Are you not using that API? In TypeScript we use those types so there's not much maintenance work -- on the C++ side we just have to use a JSON library to parse the incoming/outgoing JSON, which we haven't had much issue maintaining. I don't understand why you need to use RPC. If you're going to use RPC, you may want some tool that builds the matching TypeScript and C++ output so there's no manual copy of types.

**Reply by gnikit** [reply]
Apologies for the delayed response @sean-mcmanus I haven't had time to look into this these past couple of weeks.

So for the client side we are using the API see for example: https://github.com/fortran-lang/vscode-fortran-support/blob/f2daba191a219690befe27f416c3ca31353ebc32/src/lsp/client.ts#L134-L155 and all is fine.

For the server side currently we use use JSON (Rapid Json to be exact) to implement the entire JSON-RPC protocol, but probably this will change after this conversation. 
This is the Python version of the LS that we are replicating: https://github.com/gnikit/fortls/blob/master/fortls/langserver.py.
For example, https://github.com/gnikit/fortls/blob/0591527411e6ac07252cb83dbe5f42e703982bbf/fortls/langserver.py#L293-L295.
`request` is just a dictionary containing the LSP request in this case for [`textDocument/documentSymbol`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol).

In the Python LS I sent, I have also implemented the RPC in Python, because I thought a pure Python LS would be more portable, in hindsight it doesn't really matter. I can see how one could send the LSP response and receive the LSP requests in JSON (or Python dicts) and then construct the RPC request in Typescript.

However, doing that would still require to maintain an API of the LSP types of some sorts. For example, in the Python LS when fulfilling a `documentSymbol` request you need to send a [`DocumentSymbol`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentSymbol) (or `SymbolInformation[]` | null) which among other things requires the [`SymbolKind`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#symbolKind). In Python, we get the `SymbolKind` by [mapping](https://github.com/gnikit/fortls/blob/0591527411e6ac07252cb83dbe5f42e703982bbf/fortls/langserver.py#L274-L291) from our AST types to what effectively is the [`SymbolKind` namespace](https://github.com/microsoft/vscode-languageserver-node/blob/8b56a58d0a706bc0cc8e947b42ad135faf488fe3/types/src/main.ts#L2694-L2721) (part of the `vscode-languageserver-types`) and then package that with all the other necessary info into a [`DocumentSymbol`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentSymbol) response. So effectively, we still need to implement an LSP API of all the types used in the [`DocumentSymbol`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentSymbol) response in the server-side.
 (`SymbolKind`, `SymbolTag`, `Range`, `DocumentSymbol`, etc.)

From what I understand, in your C++ LS you have some code that produces the equivalent JSON (when stringified?) of the interfaces and namespaces found in `vscode-languageserver-types`, something similar to clangd's [Protocol.h](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/Protocol.h). Am I correct in my thinking?
We would still need to write our own version of `Protocol.h` to parse that to and from JSON, from C++ to Typescript (and vice versa), right?
There is no way we can avoid writing something similar to `Protocol.h` right?

Once again, thanks for all the help. I greatly appreciate it.

**Reply by sean-mcmanus** [reply]
Yeah, you are correct -- our cpptools has to parse/serialize the JSON it receives using some JSON library -- see https://github.com/microsoft/vscode-cpptools/blob/main/Extension/ThirdPartyNotices.txt#L2034 . We haven't found it to be a maintenance issue so we haven't investigated other approaches to potentially auto-generate the C++ code.

**Reply by gnikit** [reply]
Thanks @sean-mcmanus for all your time answering my questions, as always you've been extremely helpful!

**Reply by gnikit** [reply]
Out of curiosity, on the C++ side I suspect you keep the process running and communicate via IO buffer/network port the incoming/outgoing JSON to Typescript. You don't spin up a new process every time you request something from the server, right?

**Reply by sean-mcmanus** [reply]
Yes -- you can see that via installing our C/C++ extension and looking at the cpptools process in a task manager. The only time cpptools would get restarted is if it crashes.

**Reply by certik** [reply]
Thanks @sean-mcmanus for the clarifications. This has been extremely helpful and valuable for us to understand how to design our language server.
